File: plugins/cordova-plugin-local-notification/src/android/notification/Manager.java
/*
 * Apache 2.0 License
 *
 * Copyright (c) Sebastian Katzer 2017
 *
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apache License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://opensource.org/licenses/Apache-2.0/ and read it before using this
 * file.
 *
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 */

package de.appplant.cordova.plugin.notification;

import android.annotation.SuppressLint;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.service.notification.StatusBarNotification;
import android.support.v4.app.NotificationManagerCompat;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import de.appplant.cordova.plugin.badge.BadgeImpl;

import static android.os.Build.VERSION.SDK_INT;
import static android.os.Build.VERSION_CODES.M;
import static android.os.Build.VERSION_CODES.O;
import static android.support.v4.app.NotificationManagerCompat.IMPORTANCE_DEFAULT;
import static de.appplant.cordova.plugin.notification.Notification.PREF_KEY_ID;
import static de.appplant.cordova.plugin.notification.Notification.Type.TRIGGERED;

/**
 * Central way to access all or single local notifications set by specific
 * state like triggered or scheduled. Offers shortcut ways to schedule,
 * cancel or clear local notifications.
 */
public final class Manager {

    // TODO: temporary
    static final String CHANNEL_ID = "default-channel-id";

    // TODO: temporary
    private static final CharSequence CHANNEL_NAME = "Default channel";

    // The application context
    private Context context;

    /**
     * Constructor
     *
     * @param context Application context
     */
    private Manager(Context context) {
        this.context = context;
        createDefaultChannel();
    }

    /**
     * Static method to retrieve class instance.
     *
     * @param context Application context
     */
    public static Manager getInstance(Context context) {
        return new Manager(context);
    }

    /**
     * Check if app has local notification permission.
     */
    public boolean hasPermission () {
        return getNotCompMgr().areNotificationsEnabled();
    }

    /**
     * Schedule local notification specified by request.
     *
     * @param request Set of notification options.
     * @param receiver Receiver to handle the trigger event.
     */
    public Notification schedule (Request request, Class<?> receiver) {
        Options options    = request.getOptions();
        Notification toast = new Notification(context, options);

        toast.schedule(request, receiver);

        return toast;
    }

    /**
     * TODO: temporary
     */
    @SuppressLint("WrongConstant")
    private void createDefaultChannel() {
        NotificationManager mgr = getNotMgr();

        if (SDK_INT < O)
            return;

        NotificationChannel channel = mgr.getNotificationChannel(CHANNEL_ID);

        if (channel != null)
            return;

        channel = new NotificationChannel(
                CHANNEL_ID, CHANNEL_NAME, IMPORTANCE_DEFAULT);

        mgr.createNotificationChannel(channel);
    }

    /**
     * Update local notification specified by ID.
     *
     * @param id       The notification ID.
     * @param updates  JSON object with notification options.
     * @param receiver Receiver to handle the trigger event.
     */
    public Notification update (int id, JSONObject updates, Class<?> receiver) {
        Notification notification = get(id);

        if (notification == null)
            return null;

        notification.update(updates, receiver);

        return notification;
    }

    /**
     * Clear local notification specified by ID.
     *
     * @param id The notification ID.
     */
    public Notification clear (int id) {
        Notification toast = get(id);

        if (toast != null) {
            toast.clear();
        }

        return toast;
    }

    /**
     * Clear all local notifications.
     */
    public void clearAll () {
        List<Notification> toasts = getByType(TRIGGERED);

        for (Notification toast : toasts) {
            toast.clear();
        }

        getNotCompMgr().cancelAll();
        setBadge(0);
    }

    /**
     * Clear local notification specified by ID.
     *
     * @param id The notification ID
     */
    public Notification cancel (int id) {
        Notification toast = get(id);

        if (toast != null) {
            toast.cancel();
        }

        return toast;
    }

    /**
     * Cancel all local notifications.
     */
    public void cancelAll () {
        List<Notification> notifications = getAll();

        for (Notification notification : notifications) {
            notification.cancel();
        }

        getNotCompMgr().cancelAll();
        setBadge(0);
    }

    /**
     * All local notifications IDs.
     */
    public List<Integer> getIds() {
        Set<String> keys = getPrefs().getAll().keySet();
        List<Integer> ids = new ArrayList<Integer>();

        for (String key : keys) {
            try {
                ids.add(Integer.parseInt(key));
            } catch (NumberFormatException e) {
                e.printStackTrace();
            }
        }

        return ids;
    }

    /**
     * All local notification IDs for given type.
     *
     * @param type The notification life cycle type
     */
    public List<Integer> getIdsByType(Notification.Type type) {

        if (type == Notification.Type.ALL)
            return getIds();

        StatusBarNotification[] activeToasts = getActiveNotifications();
        List<Integer> activeIds              = new ArrayList<Integer>();

        for (StatusBarNotification toast : activeToasts) {
            activeIds.add(toast.getId());
        }

        if (type == TRIGGERED)
            return activeIds;

        List<Integer> ids = getIds();
        ids.removeAll(activeIds);

        return ids;
    }

    /**
     * List of local notifications with matching ID.
     *
     * @param ids Set of notification IDs.
     */
    private List<Notification> getByIds(List<Integer> ids) {
        List<Notification> toasts = new ArrayList<Notification>();

        for (int id : ids) {
            Notification toast = get(id);

            if (toast != null) {
                toasts.add(toast);
            }
        }

        return toasts;
    }

    /**
     * List of all local notification.
     */
    public List<Notification> getAll() {
        return getByIds(getIds());
    }

    /**
     * List of local notifications from given type.
     *
     * @param type The notification life cycle type
     */
    private List<Notification> getByType(Notification.Type type) {

        if (type == Notification.Type.ALL)
            return getAll();

        List<Integer> ids = getIdsByType(type);

        return getByIds(ids);
    }

    /**
     * List of properties from all local notifications.
     */
    public List<JSONObject> getOptions() {
        return getOptionsById(getIds());
    }

    /**
     * List of properties from local notifications with matching ID.
     *
     * @param ids Set of notification IDs
     */
    public List<JSONObject> getOptionsById(List<Integer> ids) {
        List<JSONObject> toasts = new ArrayList<JSONObject>();

        for (int id : ids) {
            Options options = getOptions(id);

            if (options != null) {
                toasts.add(options.getDict());
            }
        }

        return toasts;
    }

    /**
     * List of properties from all local notifications from given type.
     *
     * @param type
     *      The notification life cycle type
     */
    public List<JSONObject> getOptionsByType(Notification.Type type) {
        ArrayList<JSONObject> options = new ArrayList<JSONObject>();
        List<Notification> notifications = getByType(type);

        for (Notification notification : notifications) {
            options.add(notification.getOptions().getDict());
        }

        return options;
    }

    /**
     * Get local notification options.
     *
     * @param id Notification ID.
     *
     * @return null if could not found.
     */
    public Options getOptions(int id) {
        SharedPreferences prefs = getPrefs();
        String toastId          = Integer.toString(id);

        if (!prefs.contains(toastId))
            return null;

        try {
            String json     = prefs.getString(toastId, null);
            JSONObject dict = new JSONObject(json);

            return new Options(context, dict);
        } catch (JSONException e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * Get existent local notification.
     *
     * @param id Notification ID.
     *
     * @return null if could not found.
     */
    public Notification get(int id) {
        Options options = getOptions(id);

        if (options == null)
            return null;

        return new Notification(context, options);
    }

    /**
     * Set the badge number of the app icon.
     *
     * @param badge The badge number.
     */
    public void setBadge (int badge) {
        if (badge == 0) {
            new BadgeImpl(context).clearBadge();
        } else {
            new BadgeImpl(context).setBadge(badge);
        }
    }

    /**
     * Get all active status bar notifications.
     */
    StatusBarNotification[] getActiveNotifications() {
        if (SDK_INT >= M) {
            return getNotMgr().getActiveNotifications();
        } else {
            return new StatusBarNotification[0];
        }
    }

    /**
     * Shared private preferences for the application.
     */
    private SharedPreferences getPrefs () {
        return context.getSharedPreferences(PREF_KEY_ID, Context.MODE_PRIVATE);
    }

    /**
     * Notification manager for the application.
     */
    private NotificationManager getNotMgr() {
        return (NotificationManager) context.getSystemService(
                Context.NOTIFICATION_SERVICE);
    }

    /**
     * Notification compat manager for the application.
     */
    private NotificationManagerCompat getNotCompMgr() {
        return NotificationManagerCompat.from(context);
    }

}

File: plugins/cordova-plugin-local-notification/src/windows/LocalNotificationProxy/LocalNotificationProxy/LocalNotification/Manager.cs
/*
 * Apache 2.0 License
 *
 * Copyright (c) Sebastian Katzer 2017
 *
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apache License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://opensource.org/licenses/Apache-2.0/ and read it before using this
 * file.
 *
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 */

namespace LocalNotificationProxy.LocalNotification
{
    using System.Collections.Generic;
    using System.Runtime.InteropServices.WindowsRuntime;
    using Windows.UI.Notifications;

    internal class Manager
    {
        /// <summary>
        /// Gets a value indicating whether the permission to schedule notifications is enabled.
        /// </summary>
        /// <returns>True if settings are enabled</returns>
        public bool Enabled => ToastNotifier.Setting == NotificationSetting.Enabled;

        /// <summary>
        /// Gets the default toast notifier.
        /// </summary>
        internal static ToastNotifier ToastNotifier => ToastNotificationManager.CreateToastNotifier();

        /// <summary>
        /// Schedule notifications.
        /// </summary>
        /// <param name="notifications">List of key-value properties</param>
        public void Schedule([ReadOnlyArray] Options[] notifications)
        {
            foreach (Options options in notifications)
            {
                var request = new Request(options);
                var builder = new Builder(request);

                do
                {
                    var toast = builder.Build();

                    if (toast == null)
                    {
                        break;
                    }

                    ToastNotifier.AddToSchedule(toast);
                }
                while (request.MoveNext());
            }
        }

        /// <summary>
        /// Update notifications.
        /// </summary>
        /// <param name="notifications">List of key-value properties</param>
        public void Update([ReadOnlyArray] Options[] notifications)
        {
            foreach (Options options in notifications)
            {
                var bar = options.ProgressBar;

                if (!bar.Enabled)
                {
                    continue;
                }

                var data = new NotificationData { SequenceNumber = 0 };

                data.Values["progress.value"] = bar.Value.ToString();
                data.Values["progress.status"] = bar.Status.ToString();
                data.Values["progress.description"] = bar.Description.ToString();

                ToastNotifier.Update(data, options.Id.ToString());
            }
        }

        /// <summary>
        /// Clear the notifications specified by id.
        /// </summary>
        /// <param name="ids">The IDs of the notification to clear.</param>
        /// <returns>The cleared notifications.</returns>
        public List<Options> Clear(int[] ids)
        {
            var triggered = ToastNotificationManager.History.GetHistory();
            var tags = new List<int>(ids);
            var toasts = new List<Options>();

            foreach (var toast in triggered)
            {
                var id = int.Parse(toast.Tag);

                if (tags.Contains(id))
                {
                    tags.Remove(id);
                    toasts.Add(new Notification(toast).Options);
                    ToastNotificationManager.History.Remove(toast.Tag);
                }

                if (tags.Count == 0)
                {
                    break;
                }
            }

            return toasts;
        }

        /// <summary>
        /// Clear all notifications.
        /// </summary>
        public void ClearAll()
        {
            ToastNotificationManager.History.Clear();
        }

        /// <summary>
        /// Cancel the notifications specified by id.
        /// </summary>
        /// <param name="ids">The IDs of the notification to clear.</param>
        /// <returns>The cleared notifications.</returns>
        public List<Options> Cancel(int[] ids)
        {
            var scheduled = ToastNotifier.GetScheduledToastNotifications();
            var toRemove = new List<int>(ids);
            var toClear = new List<int>(ids);
            var toasts = new List<Options>();

            foreach (var toast in scheduled)
            {
                var id = int.Parse(toast.Tag);

                if (toRemove.Contains(id))
                {
                    toClear.Remove(id);
                    ToastNotifier.RemoveFromSchedule(toast);

                    if (!this.IsPhantom(toast))
                    {
                        toasts.Add(new Notification(toast).Options);
                    }
                }
            }

            if (toClear.Count > 0)
            {
                toasts.AddRange(this.Clear(toClear.ToArray()));
            }

            return toasts;
        }

        /// <summary>
        /// Cancel all notifications.
        /// </summary>
        public void CancelAll()
        {
            var toasts = ToastNotifier.GetScheduledToastNotifications();

            foreach (var toast in toasts)
            {
                ToastNotifier.RemoveFromSchedule(toast);
            }

            this.ClearAll();
        }

        /// <summary>
        /// Gets all notifications by id.
        /// </summary>
        /// <returns>List of ids</returns>
        public List<int> GetIds()
        {
            return this.GetIdsByType(Notification.Type.All);
        }

        /// <summary>
        /// Gets all notifications by id for given type.
        /// </summary>
        /// <param name="type">The type of notification.</param>
        /// <returns>List of ids</returns>
        public List<int> GetIdsByType(Notification.Type type)
        {
            var ids = new List<int>();

            if (type == Notification.Type.All || type == Notification.Type.Scheduled)
            {
                var toasts = ToastNotifier.GetScheduledToastNotifications();

                foreach (var toast in toasts)
                {
                    if (!this.IsPhantom(toast))
                    {
                        ids.Add(int.Parse(toast.Tag));
                    }
                }
            }

            if (type == Notification.Type.All || type == Notification.Type.Triggered)
            {
                var toasts = ToastNotificationManager.History.GetHistory();

                foreach (var toast in toasts)
                {
                    if (!this.IsPhantom(toast))
                    {
                        ids.Add(int.Parse(toast.Tag));
                    }
                }
            }

            return ids;
        }

        /// <summary>
        /// Gets all notifications.
        /// </summary>
        /// <returns>A list of all triggered and scheduled notifications.</returns>
        public List<Notification> GetAll()
        {
            return this.GetByType(Notification.Type.All);
        }

        /// <summary>
        /// Gets all notifications of given type.
        /// </summary>
        /// <param name="type">The type of notification.</param>
        /// <returns>A list of notifications.</returns>
        public List<Notification> GetByType(Notification.Type type)
        {
            var notifications = new List<Notification>();

            if (type == Notification.Type.All || type == Notification.Type.Scheduled)
            {
                var toasts = ToastNotifier.GetScheduledToastNotifications();

                foreach (var toast in toasts)
                {
                    if (!this.IsPhantom(toast))
                    {
                        notifications.Add(new Notification(toast));
                    }
                }
            }

            if (type == Notification.Type.All || type == Notification.Type.Triggered)
            {
                var toasts = ToastNotificationManager.History.GetHistory();

                foreach (var toast in toasts)
                {
                    if (!this.IsPhantom(toast))
                    {
                        notifications.Add(new Notification(toast));
                    }
                }
            }

            return notifications;
        }

        /// <summary>
        /// Gets all notifications.
        /// </summary>
        /// <returns>A list of notification options instances.</returns>
        public List<Options> GetOptions()
        {
            return this.GetOptionsByType(Notification.Type.All);
        }

        /// <summary>
        /// Gets notifications specified by ID.
        /// </summary>
        /// <param name="ids">Optional list of IDs to find.</param>
        /// <returns>A list of notification options instances.</returns>
        public List<Options> GetOptions(int[] ids)
        {
            var options = new List<Options>();

            foreach (var toast in this.Get(ids))
            {
                options.Add(toast.Options);
            }

            return options;
        }

        /// <summary>
        /// Gets all notifications for given type.
        /// </summary>
        /// <param name="type">The type of notification.</param>
        /// <returns>A list of notification options instances.</returns>
        public List<Options> GetOptionsByType(Notification.Type type)
        {
            var options = new List<Options>();
            var toasts = this.GetByType(type);

            foreach (var toast in toasts)
            {
                options.Add(toast.Options);
            }

            return options;
        }

        /// <summary>
        /// Gets the notifications specified by ID.
        /// </summary>
        /// <param name="ids">List of IDs to find.</param>
        /// <returns>List of found notifications.</returns>
        public List<Notification> Get(int[] ids)
        {
            var toasts = new List<Notification>();

            foreach (var id in ids)
            {
                var toast = this.Get(id);

                if (toast != null)
                {
                    toasts.Add(toast);
                }
            }

            return toasts;
        }

        /// <summary>
        /// Gets the notification by ID.
        /// </summary>
        /// <param name="id">The ID of the notification to find.</param>
        /// <returns>The found instance or null.</returns>
        public Notification Get(int id)
        {
            return this.Get(id.ToString());
        }

        /// <summary>
        /// Gets the notification by ID.
        /// </summary>
        /// <param name="id">The ID of the notification to find.</param>
        /// <returns>The found instance or null.</returns>
        public Notification Get(string id)
        {
            var scheduled = ToastNotifier.GetScheduledToastNotifications();

            foreach (var toast in scheduled)
            {
                if (toast.Tag == id)
                {
                    return new Notification(toast);
                }
            }

            var triggered = ToastNotificationManager.History.GetHistory();

            foreach (var toast in triggered)
            {
                if (toast.Tag == id)
                {
                    return new Notification(toast);
                }
            }

            return null;
        }

        /// <summary>
        /// Gets the type (scheduled, triggered or unknown).
        /// </summary>
        /// <param name="id">The ID of the notification to find for.</param>
        /// <returns>The type of the notification or unknown type.</returns>
        public Notification.Type GetType(string id)
        {
            var scheduled = ToastNotifier.GetScheduledToastNotifications();

            foreach (var toast in scheduled)
            {
                if (toast.Tag == id)
                {
                    return Notification.Type.Scheduled;
                }
            }

            var triggered = ToastNotificationManager.History.GetHistory();

            foreach (var toast in triggered)
            {
                if (toast.Tag == id)
                {
                    return Notification.Type.Triggered;
                }
            }

            return Notification.Type.Unknown;
        }

        /// <summary>
        /// If the specified toast is only for internal reason.
        /// </summary>
        /// <param name="toast">The toast notification to test for.</param>
        /// <returns>True if its an internal ("phantom") one.</returns>
        private bool IsPhantom(ToastNotification toast)
        {
            return toast.Group.Length > 0;
        }

        /// <summary>
        /// If the specified toast is only for internal reason.
        /// </summary>
        /// <param name="toast">The toast notification to test for.</param>
        /// <returns>True if its an internal ("phantom") one.</returns>
        private bool IsPhantom(ScheduledToastNotification toast)
        {
            return toast.Group.Length > 0;
        }
    }
}
